pkg/reconciler/pipelinerun/resources/pipelinerunstate.go

```diff
@@ -302,10 +308,10 @@ func (facts *PipelineRunFacts) IsGracefullyStopped() bool {
 
 // DAGExecutionQueue returns a list of DAG tasks which needs to be scheduled next
 func (facts *PipelineRunFacts) DAGExecutionQueue() (PipelineRunState, error) {
-	tasks := PipelineRunState{}
-	// when pipeline run is stopping, gracefully cancelled or stopped, do not schedule any new task and only
+	var tasks PipelineRunState
+	// when pipeline run is stopping, cancelled, gracefully cancelled or stopped, do not schedule any new task and only
 	// wait for all running tasks to complete and report their status
-	if !facts.IsStopping() && !facts.IsGracefullyCancelled() && !facts.IsGracefullyStopped() {
+	if !facts.IsStopping() && !facts.IsCancelled() && !facts.IsGracefullyCancelled() && !facts.IsGracefullyStopped() {
```

Comment:
I am surprised there is no functional change even after adding this check 🤔 

With this check, if the pipeline run is `cancelled`, it will not attempt to schedule any new task in addition to the check done in the beginning of reconciler:

https://github.com/tektoncd/pipeline/blob/9a2deb3e34ae252d32c017d0155b963aaca8218f/pkg/reconciler/pipelinerun/pipelinerun.go#L211-L215

Reply:
`stopping` is an internal state and not allowed to set by updating the pipelineRun status (at least not officially). I am wondering why are we checking these other states (cancelled, gracefullycancelled, and gracefullystopped) here? Should we move these checks along with the `cancelled` at the beginning of the reconciler?

Reply:
yup it is possible to have such gap in the existing test coverage and appreciate your effort of adding more tests 😍 

My understanding is we can definitely move the graceful cancellation out of this check but not `stopping`. The `stopping` state is something which can happen during the same reconcile cycle while `cancellation` cannot be determined until the next reconcile cycle from this point of execution, does it make sense?

Reply:
thanks! yeah I understand now that cancellation takes an extra reconcile cycle. I think pulling graceful cancellation out of this block and moving it to where cancellation is might take some extra refactoring (since we still need to attempt retries) so I would prefer it didn't block this PR, but I can create an issue and link it as a TODO if that would work? I tested this PR out using the retries example from [this comment](https://github.com/tektoncd/pipeline/pull/4651#issuecomment-1068147952) with non-graceful cancellation instead of graceful stopping and it appears to still work as intended with this PR.